lib/remote/jsonrpcconnection-pki.cpp

```diff
@@ -79,11 +79,59 @@ Value RequestCertificateHandler(const MessageOrigin::Ptr& origin, const Dictiona
 
 	if (signedByCA) {
 		if (IsCertUptodate(cert)) {
+			// The following analysis targets a direct peer's cert chain and makes no sense for forwarded CSRs.
+			if (cn == origin->FromClient->GetIdentity()) {
+				// Even if the leaf is up-to-date, the root may expire soon.
+				// In a regular setup where Icinga manages the PKI, there is only one CA.
+				// Icinga includes it in handshakes, let's see whether the peer needs a fresh one...
+
+				auto chain (SSL_get_peer_cert_chain(tlsConn.native_handle()));
+
+				if (chain) {
+					X509* root = nullptr;
+					auto len (sk_X509_num(chain));
+
+					for (int i = 0; i < len; ++i) {
+						auto link (sk_X509_value(chain, i));
+
+						if (!X509_NAME_cmp(X509_get_subject_name(link), X509_get_issuer_name(link))) {
+							root = link;
+						}
+					}
+
+					if (root && !IsCertUptodate(root)) {
+						auto oldEnd (X509_get_notAfter(root));
+						auto newEnd (X509_get_notAfter(cacert.get()));
+						int pday, psec;
+
+						if (oldEnd && newEnd && ASN1_TIME_diff(&pday, &psec, oldEnd, newEnd) && pday > 0) {
```

Comment:
Can this check more directly for the situation we're interested in? That would be something like the root sent by the remote matches our local CA certificate in all relevant aspects (I'm still not entirely sure what exactly needs to match apart from the public key) except that it our local CA has a later expiration date? If so, go ahead and send the newer one.

Reply:
Well, in the for loop I lookup the remote's root. And, if it's not up-to-date, I send an up-to-date one. But only (the last condition) if we actually have an up-to-date one. Otherwise I'd cause a reconnect loop.

Not quite sure what's the problem here.

Reply:
Without having looked at the current state in every detail, it looks like something that should probably work with most standard setups. But given that other things already work, and you also want to document it in #9825, it would be good the checks here for for updating the CA here actually check that it's pretty much the same CA expect for the expiry. Otherwise (again, without me having checked every detail so far), there might be a higher risk to for example replace multiple trusted CAs on a child node with a wrong CA.

Reply:
... except that in an ideal world all nodes get uptodate CAs from Ansible.

Also, I want pubkey rollover one nice day, so we can't do this.

Reply:
> Also, I want pubkey rollover one nice day, so we can't do this.

Please explain this in more detail. I don't see how that code would be any help for that. It sends back the very same certificate the remote endpoint is already using just with another certificate. So how would you use that for a CA key rollover? If this would send back a CA certificate with a different key, that would no longer match the signature in the endpoint certificate.

Reply:
The rough idea is, whenever a TBD periodic threshold is reached:

1. The (v2.X.0+) master creates a new CA, adds it to trust bundle
2. The latter is distributed over the mechanism already used here
3. After another TBD amount of time, certificates are issued using the successor CA (by v2.{X+1}.0+)

The master **DOES NOT** cross-sign the existing CA with the new one! This way there's always an unambiguous root with either of the public keys.

Reply:
That sounds like you'd want the code to push multiple CA certificates. The code currently sets `result->Set("ca", CertificateToString(cacert));` with `cacert` being of type `std::shared_ptr<X509>`, so isn't that just a single certificate?

Reply:
Absolutely! I've omitted that just not to spam implementation details of another construction area. Of course in a future **.0** version RequestCertificateHandler() will have to pick up the whole ApiListener::GetDefaultCaPath() and not just the first certificate for my plan to work.

Reply:
Let me try to summarize my current understanding of the logic in this PR: if a remote endpoint sends any self-signed certificate that expires earlier then the single CA certificate of the local node, it sends back this CA certificate (together with the remote endpoint's existing certificate that it used in the TLS handshake).

> Of course in a future **.0** version RequestCertificateHandler() will have to pick up the whole ApiListener::GetDefaultCaPath() and not just the first certificate for my plan to work.

So this PR would make it into 2.14.1 and another one for CA renewal into 2.15.0 for example? Without extra checks that we actually push a renewed version of the otherwise same CA, that sounds quite dangerous to me, as that could end up with a satellite that then starts sending out the wrong CA.

> I've omitted that just not to spam implementation details of another construction area.

It's not entirely separate if you use it as a reason why this PR should look like it does.

Reply:
OK, let's forget that theoretical construction area for a moment and get back to https://github.com/Icinga/icinga2/pull/9891#discussion_r1385170249.

* #9825

are just docs and not important if not ready. This thing here **is**. If it prevents those docs from merged as-is, I can live with that.

And RequestCertificateHandler() **already** responds with a single CA, but just if your leaf is about to expire. Here I just add another condition for the same event, except that the "renewed" leaf isn't renewed.

If we have a problem, we have it either already (issue welcome) or with #9825 (where I can conceal multiple root CAs for now if you wish) or with https://github.com/Icinga/icinga2/pull/9891#discussion_r1386279321 (which exists just on paper, yet). But we don't have a problem with the PR you're looking at right now.

Reply:
> And RequestCertificateHandler() already responds with a single CA, but just if your leaf is about to expire.

But this should only happen if there was a new certificate issues, which only happens if some node in chain (presumably the master) has the CA key.

So wouldn't the changed code might try to replace certificates that would never have been touched before. Thus, I'm suggesting a more restrictive check so that we only replace certificates where we know that we don't break something.